che-theia-factory-frontend-extension/src/browser/factory-theia-client.ts

```diff
@@ -0,0 +1,95 @@
+        @inject(EnvVariablesServer) private readonly envVariablesServer: EnvVariablesServer,
+        @inject(Git) protected readonly git: Git,
+    ) { }
+
+    onStart(app: FrontendApplication): void {
+        const factoryid = queryString.parse(window.location.search)['factory-id'];
+
+        if (!factoryid) {
+            return;
+        }
+
+        this.envVariablesServer.getVariables().then((envVariables) => {
+
+
+            const cheApiExternalVar = envVariables.find(function(envVariable) { return envVariable.name === 'CHE_API_EXTERNAL' });
+            const projectsRootEnvVar = envVariables.find(function(envVariable) { return envVariable.name === 'CHE_PROJECTS_ROOT' });
+
+
+            if (!cheApiExternalVar) {
+                return;
+            }
+
+            var projectsRoot = "/projects";
+            if (projectsRootEnvVar && projectsRootEnvVar.value) {
+                projectsRoot = projectsRootEnvVar.value;
+            }
+
+            const factory = new Resources(FactoryTheiaClient.axiosInstance, String(cheApiExternalVar.value));
+            factory.getById<IFactory>(factoryid).then((response: AxiosResponse<IFactory>) => {
+                response.data.workspace.projects.forEach((project: IProjectConfig) => {
+                    if (!project.source) {
+                        return;
+                    }
+
+                    const source = project.source;
+                    const projectPath = projectsRoot + project.path;
+
+                    this.messageService.info(`Cloning ... ${source.location} to ${projectPath}...`)
+                    console.info(`Cloning ... ${source.location} to ${projectPath}...`)
+                    setTimeout(() => {
+                        console.info(`Cloning after 10 secs... ${source.location} to ${projectPath}...`)
+                        this.git.clone(
+                            source.location,
+                            {
+                                localUri: projectPath
+                            }
+                        ).then(
+                            (repo: Repository) => {
+                                this.messageService.info(`Project ${projectPath} successfully cloned.`);
+                                if (source.parameters['branch']) {
+                                    const options: Git.Options.Checkout.Branch = { branch: source.parameters['branch'] };
+                                    this.git.checkout(repo, options);
+                                }
+                            }
+                        ).catch((error) => {
+                            console.error(`Couldn't clone ${source.location} to ${projectPath}... ${error}`);
+                            this.messageService.error(`Couldn't clone ${source.location} to ${projectPath}... ${error}`);
+                        });
+                    }, 10000);
+
+                });
+            });
+        })
+    }
```

Comment:
Can you, please, rewrite this method with [async/await](https://basarat.gitbooks.io/typescript/docs/async-await.html), and maybe split in several, it is hard to read code with so many callbacks